warehouse/slave.go

```diff
@@ -189,126 +207,170 @@ func (job *Payload) getDiscardsTable() string {
 }
 
-func (jobRun *JobRun) getLoadFilePath(tableName string) string {
-	job := jobRun.job
+func (jr *JobRun) getLoadFilePath(tableName string) string {
+	job := jr.job
 	randomness := misc.FastUUID().String()
-	return strings.TrimSuffix(jobRun.stagingFilePath, "json.gz") + tableName + fmt.Sprintf(`.%s`, randomness) + fmt.Sprintf(`.%s`, warehouseutils.GetLoadFileFormat(job.LoadFileType))
+	return strings.TrimSuffix(jr.stagingFilePath, "json.gz") + tableName + fmt.Sprintf(`.%s`, randomness) + fmt.Sprintf(`.%s`, warehouseutils.GetLoadFileFormat(job.LoadFileType))
 }
 
 func (job *Payload) getColumnName(columnName string) string {
 	return warehouseutils.ToProviderCase(job.DestinationType, columnName)
 }
 
-type loadFileUploadJob struct {
-	tableName  string
-	outputFile encoding.LoadFileWriter
-}
-
-type loadFileUploadOutput struct {
-	TableName             string
-	Location              string
-	TotalRows             int
-	ContentLength         int64
-	StagingFileID         int64
-	DestinationRevisionID string
-	UseRudderStorage      bool
-}
+// uploadLoadFiles returns the upload output for each file uploaded to object storage
+func (jr *JobRun) uploadLoadFiles(ctx context.Context) ([]loadFileUploadOutput, error) {
+	ctx, cancel := context.WithTimeout(ctx, jr.slaveUploadTimeout)
+	defer cancel()
 
-func (jobRun *JobRun) uploadLoadFilesToObjectStorage(ctx context.Context) ([]loadFileUploadOutput, error) {
-	job := jobRun.job
-	uploader, err := job.getFileManager(job.DestinationConfig, job.UseRudderStorage)
+	uploader, err := jr.job.getFileManager(jr.job.DestinationConfig, jr.job.UseRudderStorage)
 	if err != nil {
-		return []loadFileUploadOutput{}, err
-	}
-	// var loadFileIDs []int64
-	var loadFileUploadOutputs []loadFileUploadOutput
-
-	// take the first staging file id in upload
-	// TODO: support multiple staging files in one upload
-	stagingFileId := jobRun.job.StagingFileID
-
-	loadFileOutputChan := make(chan loadFileUploadOutput, len(jobRun.outputFileWritersMap))
-	loadFileUploadTimer := jobRun.timerStat("load_file_upload_time")
-	uploadJobChan := make(chan *loadFileUploadJob, len(jobRun.outputFileWritersMap))
-	// close chan to avoid memory leak ranging over it
-	defer close(uploadJobChan)
-	uploadErrorChan := make(chan error, numLoadFileUploadWorkers)
-	ctx, cancel := context.WithCancel(ctx)
-	defer cancel()
-	for i := 0; i < numLoadFileUploadWorkers; i++ {
-		go func(ctx context.Context) {
-			for uploadJob := range uploadJobChan {
+		return nil, fmt.Errorf("creating uploader: %w", err)
+	}
+
+	type result struct {
+		uploadOutput loadFileUploadOutput
+		err          error
+	}
+
+	type uploadJob struct {
+		tableName  string
+		outputFile encoding.LoadFileWriter
+	}
+
+	var output []loadFileUploadOutput
+	var totalUploadTime time.Duration
+	var mu sync.Mutex
+
+	defer func() {
+		if totalUploadTime > 0 {
+			jr.timerStat("load_file_upload_time").SendTiming(totalUploadTime)
+		}
+	}()
+
+	generate := func() <-chan *uploadJob {
```

Comment:
I don't understand why we need `generate`. Why don't you iterate over `jr.outputFileWritersMap` inside `process` itself?
You are already leveraging the `group` limit to ensure the maximum number of workers so getting rid of `generate` should be simple and it would help us decrease the complexity of the `JobRun.uploadLoadFiles` method.

Have a look at the `git diff` below. I suggest we apply all the changes from the diff.

As a sidenote (which you'll find in the diff as well) you are using a group with context but you're ignoring the context. If you don't plan to use it then you could just use a simple group or do the synchronisation natively. Since I think the group context in this case it's going to be useful, I've included it in the diff:

```diff
diff --git a/warehouse/slave.go b/warehouse/slave.go
index 06952f3de..467cb6939 100644
--- a/warehouse/slave.go
+++ b/warehouse/slave.go
@@ -247,63 +247,48 @@ func (jr *JobRun) uploadLoadFiles(ctx context.Context) ([]loadFileUploadOutput,
 		}
 	}()
 
-	generate := func() <-chan *uploadJob {
-		jobStream := make(chan *uploadJob, len(jr.outputFileWritersMap))
-		go func() {
-			defer close(jobStream)
-
-			for tableName, loadFile := range jr.outputFileWritersMap {
-				select {
-				case <-ctx.Done():
-					return
-				default:
-					jobStream <- &uploadJob{
-						tableName:  tableName,
-						outputFile: loadFile,
-					}
-				}
-			}
-		}()
-		return jobStream
-	}
-
-	uploadLoadFile := func(uploader filemanager.FileManager, uploadFile encoding.LoadFileWriter, tableName string) (filemanager.UploadedFile, error) {
+	uploadLoadFile := func(
+		ctx context.Context, uploader filemanager.FileManager, uploadFile encoding.LoadFileWriter, tableName string,
+	) (filemanager.UploadedFile, error) {
 		file, err := os.Open(uploadFile.GetLoadFile().Name())
 		if err != nil {
-			return filemanager.UploadedFile{}, fmt.Errorf("opening file: %w", err)
+			return filemanager.UploadedFile{}, fmt.Errorf("opening file %s: %w", uploadFile.GetLoadFile().Name(), err)
 		}
-		defer file.Close()
+		defer func() { _ = file.Close() }()
 
-		var uploadLocation filemanager.UploadedFile
 		if slices.Contains(warehouseutils.TimeWindowDestinations, jr.job.DestinationType) {
-			uploadLocation, err = uploader.Upload(
+			return uploader.Upload(
 				ctx,
 				file,
 				warehouseutils.GetTablePathInObjectStorage(jr.job.DestinationNamespace, tableName),
 				jr.job.LoadFilePrefix,
 			)
-		} else {
-			uploadLocation, err = uploader.Upload(
-				ctx,
-				file,
-				jr.loadObjectFolder,
-				tableName,
-				jr.job.SourceID,
-				getBucketFolder(jr.job.UniqueLoadGenID, tableName),
-			)
 		}
-		return uploadLocation, err
+		return uploader.Upload(
+			ctx,
+			file,
+			jr.loadObjectFolder,
+			tableName,
+			jr.job.SourceID,
+			getBucketFolder(jr.job.UniqueLoadGenID, tableName),
+		)
 	}
 
-	process := func(uploadJobChan <-chan *uploadJob) <-chan *result {
+	process := func() <-chan *result {
 		processStream := make(chan *result, len(jr.outputFileWritersMap))
 
-		g, _ := errgroup.WithContext(ctx)
+		g, groupCtx := errgroup.WithContext(ctx)
 		g.SetLimit(jr.numLoadFileUploadWorkers)
 
-		for i := 0; i < jr.numLoadFileUploadWorkers; i++ {
-			g.Go(func() error {
-				for uploadJob := range uploadJobChan {
+		go func() {
+			defer close(processStream)
+
+			for tableName, loadFile := range jr.outputFileWritersMap {
+				uploadJob := &uploadJob{
+					tableName:  tableName,
+					outputFile: loadFile,
+				}
+				g.Go(func() error {
 					select {
 					case <-ctx.Done():
 						return ctx.Err()
@@ -311,6 +296,7 @@ func (jr *JobRun) uploadLoadFiles(ctx context.Context) ([]loadFileUploadOutput,
 						loadFileUploadStart := time.Now()
 
 						uploadOutput, err := uploadLoadFile(
+							groupCtx,
 							uploader,
 							uploadJob.outputFile,
 							uploadJob.tableName,
@@ -340,22 +326,19 @@ func (jr *JobRun) uploadLoadFiles(ctx context.Context) ([]loadFileUploadOutput,
 							},
 						}
 					}
-				}
-				return nil
-			})
-		}
+					return nil
+				})
+			}
 
-		go func() {
 			if err := g.Wait(); err != nil {
 				processStream <- &result{err: err}
 			}
-			close(processStream)
 		}()
 
 		return processStream
 	}
 
-	for processedJob := range process(generate()) {
+	for processedJob := range process() {
 		if err := processedJob.err; err != nil {
 			return nil, fmt.Errorf("uploading load file to object storage: %w", err)
 		}
```

As far as I can see the tests pass, please have a look :+1:

Reply:
I was overcomplicating it using `generate` since it is doing nothing just adding the `uploadJob` to the channel which can be done in `process` as well.